File: #124_Binary_Tree_Maximum_Path_Sum /Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
	int maxSum = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        if (root == null) return 0;
        // int maxSum = 0;
        findSum(root);
        return maxSum;
    }

    private int findSum(TreeNode node) {
    	if (node == null) return 0;
    	int right = findSum(node.right);
    	int left = findSum(node.left);
    	if (right < 0) right = 0;
    	if (left < 0) left = 0;
    	if (right + left + node.val > maxSum) maxSum = right + left + node.val;
    	return Math.max(left, right) + node.val;
    }
}

File: #124_Binary_Tree_Maximum_Path_Sum /Solution.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:

    def __init__(self):
        self.ans = 0

    # @param {TreeNode} root
    # @return {integer}
    def maxPathSum(self, root):
        self.ans = 0
        self.maxBranch(root)
        return self.ans

    def maxBranch(self, root):
        if not root:
            return 0
        ld = max(0, self.maxBranch(root.left))
        rd = max(0, self.maxBranch(root.right))
        self.ans = max(self.ans, ld + rd + root.val)
        return max(ld, rd) + root.val

sol = Solution()
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(-15)
print sol.maxPathSum(root)